### PR TITLE
CORS configuration support

### PR DESCRIPTION
**Pull request by sumeetchhetri**
Set server port value in ^%webhttp
Add support for CORS configuration (Refer - https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#The_HTTP_response_headers)
Set cors configuration in ^%webhttp
Updated unit test as well

### CODE REVIEW

**Comment by sumeetchhetri** [comment]
@shabiel, please review, once this is merged, I will start creating additional issues and use-cases/real-life-examples supporting the new features along with PR's for the same.

**Comment by shabiel** [comment]
This is much better!

I think because we try to avoid to use globals, can you re-write the
CORS support so that you pass it in to the web server when starting as
a parameter rather than getting the configuration information from
globals?

Plus, the test you supplied doesn't do anything. Can you make sure it
makes a call to exercise your code? Look at the CORS test currently
there.

--Sam

**Comment by sumeetchhetri** [comment]
Sure Sam, if you look at the code (https://github.com/shabiel/M-Web-Server/pull/37/files#diff-cc22c80b61b1d84e543a9606e629cd23) you will see that I have added a new setupCORS function to add configuration for CORS, which in turn is used in by the flow in the _webrsp.m file (the global check)

**Comment by shabiel** [comment]
Sumeet,

I hope you understood me. The Unit Test needs to run your main code, in
addition to setting up CORS.

--Sam

**Comment by sumeetchhetri** [comment]
You are right Sam, the test that you have already written for CORS will only pass when my setupCORS function is called from the STARTUP function. So I did not have to write any new test case for the same. What I can do is I can write another test case for CORS where I do not setupCORS config and then check that I don't get any CORS headers back. Is this what you want?

**Comment by shabiel** [comment]
Yes. That's it.

Also, I did mention just a moment ago that I won't take the CORS work
as is. You need to re-write it so that it can be configured by sending
parameters to the start-up to the server (start^%webreq) rather than
via global configuration.

--Sam

**Comment by sumeetchhetri** [comment]
I have made the necessary changes required to pass in the CORS configuration parameters to the routine in the NOGBL mode, please verify accordingly.

**Comment by shabiel** [comment]
That's good. We are moving in the right direction!

Did you run the test, and did it pass?

So let's make these two further changes:
1. Instead of using CORSENAB,CORSHDRS,CORSMETH,CORSORG,CORSCRED, just
have a single CORS variable that is passed by reference.
2. Eliminate the use of the global. I.e., everybody should use the
code for NOGBL mode.

--Sam

**Comment by sumeetchhetri** [comment]
Yeah that was the first thing I did, but then I got errors while passing the array to the jobs (pass by reference error) and hence converted them to individual variables. Also for some weird reason or the other the tests get blocked after the first 3-4 tests on my Vista/GT.M installation, not sure why that happens though.

**Comment by shabiel** [comment]
You are right. You cannot pass reference variables to jobbed off processes.

Why don't you do this first, and then we can work to refine it
together? Keep it as is, but eliminate the use of the global.

**Comment by sumeetchhetri** [comment]
Sure Sam

**Comment by sumeetchhetri** [comment]
Done with the suggested changes

**Comment by shabiel** [comment]
Okay. Thanks. Let me ask Christopher to give me his opinion of how
things are going so far.

By the way, we are off on holiday for the next 3 days, so you won't
get any responses from me.

File: src/_webrsp.m [hunk 1]
```diff
@@ -273,10 +273,13 @@ N ARGS S ARGS("*")=$E(HTTPREQ("path"),2,9999)
...
- I $G(HTTPREQ("method"))="OPTIONS" D W("Access-Control-Allow-Methods: OPTIONS, POST"_$C(13,10))
- I $G(HTTPREQ("method"))="OPTIONS" D W("Access-Control-Allow-Headers: Content-Type"_$C(13,10))
- I $G(HTTPREQ("method"))="OPTIONS" D W("Access-Control-Max-Age: 86400"_$C(13,10))
- D W("Access-Control-Allow-Origin: *"_$C(13,10))
+ I $G(CORS("enabled"))="Y" D
+ . I $G(HTTPREQ("method"))="OPTIONS" D
+ . . D W("Access-Control-Allow-Credentials: "_CORS("credentials")_$C(13,10))
+ . . D W("Access-Control-Allow-Methods: "_CORS("method")_$C(13,10))
+ . . D W("Access-Control-Allow-Headers: "_CORS("header")_$C(13,10))
+ . . D W("Access-Control-Max-Age: 86400"_$C(13,10))
```
Comment (ChristopherEdwards) [inline src/_webrsp.m]: this should probably be an option as well.

Comment (sumeetchhetri) [reply src/_webrsp.m]: Done

**Comment by sumeetchhetri** [comment]
Have updated the documentation as requested.

**Comment by shabiel** [comment]
I need to check with Christopher.

I hope not to discourage you, but right now M-Web-Server is not a
priority for me at my job right now; so I can't pay much attention to
new features.

If you want something to do, the documentation really needs to be updated.

--Sam

**Comment by sumeetchhetri** [comment]
Thats good to hear @shabiel, this gives me some direction on how to move forward then. We are going to use M-Web-Server as a very important component of our Vista based EMR deployments. I see a lot of improvements in terms of existing code base for this project and a lot of inherent issues like JSON serialization which I have already fixed in our internal code-base for M-Web-Server. Henceforth I would choose to maintain all these configuration changes, bug fixes and custom changes/improvements in my own forked repository. Thanks for your time anyways.

**Comment by shabiel** [comment]
Sumeet,

Thank you for understanding. I hope you will keep your source code
open source and let us know of features that you add.

--Sam

**Pull request closed without merging**